File: src/main/java/top/chenray/antilitematica/util/TokenBucket.java
package top.chenray.antilitematica.util;

public final class TokenBucket {
   private final double refillPerMillis;
   private final double capacity;
   private double tokens;
   private long lastRefillMs;

   private TokenBucket(double refillPerSecond, double capacity) {
      this.refillPerMillis = refillPerSecond / (double)1000.0F;
      this.capacity = Math.max((double)1.0F, capacity);
      this.tokens = this.capacity;
      this.lastRefillMs = System.currentTimeMillis();
   }

   public static TokenBucket perSecond(double rate, double capacity) {
      return new TokenBucket(rate, capacity);
   }

   public boolean tryConsume(int amount) {
      if (amount <= 0) return true;
      this.refill();
      if (this.tokens + 1.0E-9 < (double)amount) return false;
      this.tokens -= (double)amount;
      return true;
   }

   private void refill() {
      long now = System.currentTimeMillis();
      long delta = now - this.lastRefillMs;
      if (delta > 0L) {
         this.tokens = Math.min(this.capacity, this.tokens + (double)delta * this.refillPerMillis);
         this.lastRefillMs = now;
      }
   }
}

File: src/main/kotlin/icu/epochcraft/antilitematica/util/TokenBucket.kt
package icu.epochcraft.antilitematica.util

/**
 * 令牌桶限速器：控制单位时间内的操作次数（防 Printer 放置限速用）。
 *
 * @author 阿清
 */
class TokenBucket(private val capacity: Int, private val refillPerSecond: Int) {

    private var tokens: Double = capacity.toDouble()
    private var lastRefill: Long = System.currentTimeMillis()

    /** 尝试消耗 [count] 个令牌，不足返回 false */
    @Synchronized
    fun tryConsume(count: Int): Boolean {
        refill()
        if (tokens >= count) {
            tokens -= count
            return true
        }
        return false
    }

    private fun refill() {
        val now = System.currentTimeMillis()
        val elapsedSeconds = (now - lastRefill) / 1000.0
        if (elapsedSeconds > 0) {
            tokens = minOf(capacity.toDouble(), tokens + elapsedSeconds * refillPerSecond)
            lastRefill = now
        }
    }

    companion object {
        /** 按每秒速率创建（容量 = 突发余量） */
        fun perSecond(rate: Int, capacity: Int = rate * 2): TokenBucket = TokenBucket(capacity, rate)
    }
}
